internal/api_client/auth/snowflake/token_provider.go

```
@@ -0,0 +1,235 @@
+package snowflake
+
+// Copyright (C) 2025 by Posit Software, PBC.
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/golang-jwt/jwt/v5"
+)
+
+// TokenProvider uses a snowflake connection to get an ingress access token.
+type TokenProvider interface {
+	// GetToken returns an access token for the given host
+	GetToken(host string) (string, error)
+}
+
+// JWTTokenProvider uses key-pair authentication.
+type JWTTokenProvider struct {
+	account    string
+	user       string
+	privateKey *rsa.PrivateKey
+	role       string // Optional role to use for the token
+
+	tokenEndpoint string // OAuth token exchange URL, to override in tests
+}
+
+var _ TokenProvider = &JWTTokenProvider{}
+
+// NewJWTTokenProvider creates a new TokenProvider that uses JWT authentication,
+// loading the private key from the specified file path
+func NewJWTTokenProvider(
+	account string,
+	user string,
+	privateKeyFile string,
+	role string,
+) (*JWTTokenProvider, error) {
+	pemData, err := os.ReadFile(privateKeyFile)
+	if err != nil {
+		return nil, fmt.Errorf("error loading private key file: %w", err)
+	}
+
+	block, _ := pem.Decode(pemData)
```

Is this an ignored error?